types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,658 @@
+	struct wlr_drm_lease_v1 *lease = drm_lease_v1_from_resource(resource);
+	drm_lease_v1_destroy(lease);
+}
+
+static void drm_lease_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_v1_interface lease_impl = {
+	.destroy = drm_lease_v1_handle_destroy,
+};
+
+static void drm_lease_request_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(resource);
+	drm_lease_request_v1_destroy(request);
+}
+
+static void drm_lease_request_v1_handle_request_connector(
+		struct wl_client *client, struct wl_resource *request_resource,
+		struct wl_resource *connector_resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(request_resource);
+	struct wlr_drm_lease_connector_v1 *connector =
+		drm_lease_connector_v1_from_resource(connector_resource);
+
+	if (!connector || connector->active_lease) {
+		/* This connector offer has been withdrawn or is leased */
+		wlr_log(WLR_ERROR, "Failed to request connector");
+		request->invalid = true;
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Requesting connector %s", connector->output->name);
+
+	size_t n_connectors = request->n_connectors + 1;
+	request->connectors = realloc(request->connectors,
+			n_connectors * sizeof(struct wlr_drm_lease_connector_v1*));
+	request->connectors[request->n_connectors] = connector;
+	request->n_connectors = n_connectors;
+
+	/* Connector is not available for other clients */
+	struct wl_resource *resource, *tmp;
+	wl_resource_for_each_safe(resource, tmp, &connector->resources) {
+		if (resource != connector_resource) {
+			wp_drm_lease_connector_v1_send_withdrawn(resource);
+		}
+	}
+}
+
+static void drm_lease_request_v1_handle_submit(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_request_v1 *request =
+			drm_lease_request_v1_from_resource(resource);
+
+	struct wlr_drm_lease_v1 *lease = calloc(1, sizeof(struct wlr_drm_lease_v1));
+	if (!lease) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *lease_resource = wl_resource_create(client,
+			&wp_drm_lease_v1_interface, version, id);
+	if (!lease_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(lease);
+		return;
+	}
+
+	lease->device = request->device;
+	lease->resource = lease_resource;
+	wl_list_insert(&lease->device->leases,
+			wl_resource_get_link(lease_resource));
+
+	request->lease = lease;
+
+	wl_resource_set_implementation(lease_resource, &lease_impl,
+			lease, drm_lease_v1_handle_resource_destroy);
+
+	if (request->invalid) {
+		/* Pre-emptively reject invalid lease requests */
+		wp_drm_lease_v1_send_finished(lease->resource);
```

Comment:
This sends the `finished` event to the lease, but doesn't destroy the `wlr_drm_lease_v1`.